src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,562 @@
+        tabular.record('{}/dLoss'.format(self.policy.name),
+                       loss_before - loss_after)
+        tabular.record('{}/KLBefore'.format(self.policy.name),
+                       policy_kl_before)
+        tabular.record('{}/KL'.format(self.policy.name), policy_kl)
+        pol_ent = self._f_policy_entropy(*policy_opt_input_values)
+        ent = np.sum(pol_ent) / np.sum(samples_data['valids'])
+        tabular.record('{}/Entropy'.format(self.policy.name), ent)
+        tabular.record('{}/Perplexity'.format(self.policy.name), np.exp(ent))
+
+        self._fit_baseline_with_data(samples_data)
+
+        ev = np_tensor_utils.explained_variance_1d(samples_data['baselines'],
+                                                   samples_data['returns'],
+                                                   samples_data['valids'])
+
+        tabular.record('{}/ExplainedVariance'.format(self.baseline.name), ev)
+        self._old_policy.model.parameters = self.policy.model.parameters
+
+    def _build_inputs(self):
+        """Build input variables.
+
+        Returns:
+            namedtuple: Collection of variables to compute policy loss.
+            namedtuple: Collection of variables to do policy optimization.
+
+        """
+        observation_space = self.policy.observation_space
+        action_space = self.policy.action_space
+
+        with tf.name_scope('inputs'):
+            if self.flatten_input:
+                obs_var = tf.compat.v1.placeholder(
+                    tf.float32,
+                    shape=[None, None, observation_space.flat_dim],
+                    name='obs')
+            else:
+                obs_var = observation_space.to_tf_placeholder(name='obs',
+                                                              batch_dims=2)
+            action_var = action_space.to_tf_placeholder(name='action',
+                                                        batch_dims=2)
+            reward_var = new_tensor(name='reward', ndim=2, dtype=tf.float32)
+            valid_var = tf.compat.v1.placeholder(tf.float32,
+                                                 shape=[None, None],
+                                                 name='valid')
+            length_var = new_tensor(name='length', ndim=1, dtype=tf.int32)
+            baseline_var = new_tensor(name='baseline',
+                                      ndim=2,
+                                      dtype=tf.float32)
+
+            policy_state_info_vars = {
+                k: tf.compat.v1.placeholder(tf.float32,
+                                            shape=[None] * 2 + list(shape),
+                                            name=k)
+                for k, shape in self.policy.state_info_specs
+            }
+            policy_state_info_vars_list = [
+                policy_state_info_vars[k] for k in self.policy.state_info_keys
+            ]
+
+        if self.policy.recurrent:
```

Comment:
can you remove this from the new policies and just _always_ augment with extra_state_var? if a policy is stateless, it can just have `len(state_info_keys) == 0`

Reply:
This flag is for applying `boolean_mask` or not. For recurrent, inputs feeding to policies will be [N, T, S^*] while non-recurrent would be (N, S^*)

Reply:
is there a reason they both can't take the same feed?

Reply:
They both take the same feed -- (N, T, S^), and for non-recurrent policies, since

> Can you repost this as modifying, rather than renaming, NPO/PPO/TRPO? Alternatively, if you're not going to do that in this PR, can you send us a branch which does that?
> 
> This PR makes it difficult to compare your changes between the two versions of the algorithms.

That makes sense. I will repost it.

Reply:
Sampler provides trajectories (N, T, S^) to the algorithm, and then passed to the policies. For recurrent policies, the shape matches. For non-recurrent policies, the shape has to be (N, S^) unless we augment it to be (N, 1, S^) to match the shape.

Reply:
I think it's part of the API in garage that we need to clean up and unify -- Whether we use dense matrix for both recurrent and non-recurrent policies.

Reply:
how hard is it to make the non-recurrent policies accept `(N, T, S^)`? -- currently this loss function already computes assuming the inputs are `(N, T, S^)`

Reply:
Indeed it computes advantage assuming inputs are `(N, T, S^)`, but then filter with valids to make it `(N * [T], S^)`. To make non-recurrent policies accept `(N, T, S^)`, we basically feed a placeholder with such shape, which means whenever the policy accepts data (observation/actions), we need to augment it from `(N, S^)` to `(N, 1, S^)`.

Reply:
Okay -- please look at it. If it's too much work, please file an issue about doing this before submitting.

Reply:
Regarding augmentation with extra_state_var -- What if a recurrent policy with `state_include_action=False`?

Reply:
I realized in that case `policy_state_info_list` will be empty (since `extra_state_var` is not added if `state_include_action=False`)